snake.asm

```diff
@@ -22,17 +23,17 @@ start:                  ; reset game
     aam 0x14            ;     IMUL (AH is irrelevant here), AAM and AAD with some magic constants maps up => -80, left => -2, right => 2, down => 80
     aad 0x44            ;     using arithmetic instructions is more compact than checks and conditional jumps
     cbw                 ;     but causes weird snake movements though with other keys
-    add di, ax          ; add offset to head position
-    cmp di, bx          ; check if head crossed vertical edge by comparing against screen size in BX
-    lodsw               ; load 0x2007 into AX from off-screen screen buffer and advance head pointer
-    adc [di], ah        ; ADC head position with 0x20 to set snake character
-    jnp start           ;   if it already had snake or wall in it or if it crossed a vertical edge, PF=0 from ADC => game over
-    mov [bp+si], di     ; store head position, use BP+SI to default to SS
-    jz .food            ; if food was consumed, ZF=1 from ADC => generate new food
+    add ax, si          ; set AX to new head position
+    cmp bx, ax          ; check if head crossed vertical edge by comparing against BX
+    stosw               ; store head position (SS=ES by default) and advance head pointer
+    xchg ax, si         ; save new head position to SI
+    sbb [si], bl        ; SBB head position with 0xD0 to set snake character
+    jnp start           ;   if it already had snake or wall in it or if it crossed a vertical edge, PF=0 from SBB => game over
+    ja .food            ; if food was consumed, CF=0 from SBB => generate new food
 .wall:                  ; draw an invisible wall on the left side
-    mov [bx], cl        ;   store wall character
     sub bx, BYTE 0x50   ;   go one line backwards
+    mov [bx], cl        ;   store wall character
     jns .wall           ; jump to draw the next wall
     pop bx              ; no food was consumed so pop tail position into BX
-    mov [bx], ah        ; clear old tail position on screen
+    mov [bx], BYTE 0x20 ; clear old tail position on screen

```

Comment:
And also this, as any position after `0x7D0` is guaranteed to contain `0x20` but we can't use it either